## 7.1 The MultiIndex Object

MuliIndex는 관계형db에서 다중 key와 같은 역할을 한다.

In [1]:
import pandas as pd

In [2]:
address = ("8809 Flair Square", "Toddside", "IL", "37206")
address

('8809 Flair Square', 'Toddside', 'IL', '37206')

In [3]:
addresses = [
    ("8809 Flair Square", "Toddside", "IL", "37206"),
    ("9901 Austin Street", "Toddside", "IL", "37206"),
    ("905 Hogan Quarter", "Franklin", "IL", "37206"),
]

In [4]:
#pandas내 MultiIndex class에는 from_tuples라는 class가 있다.
#이는 tuple list로부터 MultiIndex를 instance화할 수 있다.
#다음은 addresses list를 넘겨 from_tuples class method를 호출한다.


# The two lines below are equivalent
pd.MultiIndex.from_tuples(addresses)
pd.MultiIndex.from_tuples(tuples = addresses)

MultiIndex([( '8809 Flair Square', 'Toddside', 'IL', '37206'),
            ('9901 Austin Street', 'Toddside', 'IL', '37206'),
            ( '905 Hogan Quarter', 'Franklin', 'IL', '37206')],
           )

위 코드의 결과로 4개의 요소(level)로 구성된 3개의 tuple을 저장하는 MultiIndex를 생성하였다.

예를 들어'8809 Flair Square', '9901 Austin Street', '905 Hogan Quarter'이 같은 level에 속한다.

In [5]:
#from_tuples()의 names 매개변수에 list를 전달하면 각 MultiIndex level에 이름을 할당할 수 있다.
row_index = pd.MultiIndex.from_tuples(
    tuples = addresses,
    names = ["Street", "City", "State", "Zip"]  
)

row_index

MultiIndex([( '8809 Flair Square', 'Toddside', 'IL', '37206'),
            ('9901 Austin Street', 'Toddside', 'IL', '37206'),
            ( '905 Hogan Quarter', 'Franklin', 'IL', '37206')],
           names=['Street', 'City', 'State', 'Zip'])


MultiIndex는 각 label이 여러 값을 보유하는 container이다. level은 label에서 동일한 위치에 있는 값으로 구성된다.


In [6]:
#위 row_index를 사용하기 위해서는 3개의 행이 있는 dataset이 필요하다.
data = [
    ["A", "B+"],
    ["C+", "C"],
    ["D-", "A"],
]

columns = ["Schools", "Cost of Living"]

#DataFrame의 index 매개변수에 MultiIndex를 전달하여 MultiIndex를 DataFrame에 연결할 수 있다.
area_grades = pd.DataFrame(
    data = data, index = row_index, columns = columns
)

area_grades

,,,,Schools,Cost of Living
Street,City,State,Zip,,
8809 Flair Square,Toddside,IL,37206,A,B+
9901 Austin Street,Toddside,IL,37206,C+,C
905 Hogan Quarter,Franklin,IL,37206,D-,A


위 DataFrame의 index는 MultiIndex이므로, 각 행은 MultiIndex의 각 요소로 구분된다.

In [7]:
area_grades.columns

Index(['Schools', 'Cost of Living'], dtype='object')

In [8]:
column_index = pd.MultiIndex.from_tuples([
    ("Culture", "Restaurants"),
    ("Culture", "Museums"),
    ("Services", "Police"),
    ("Services", "Schools"),
])

column_index

MultiIndex([( 'Culture', 'Restaurants'),
            ( 'Culture',     'Museums'),
            ('Services',      'Police'),
            ('Services',     'Schools')],
           )

In [9]:
#위 열 축(column_index)에 대한 MultiIndex를 사용하기 위해서는 4개의 열이 있는 dataset이 필요하다.
data = [
    ["C-", "B+", "B-", "A"],
    ["D+", "C", "A", "C+"],
    ["A-", "A", "D+", "F"], 
]

In [10]:
#행과 열 축 모두 MultiIndex를 사용하여 DataFrame을 생성할 수 있다.
pd.DataFrame(
    data = data, index = row_index, columns = column_index
)

Culture         Services        
                                        Restaurants Museums   Police Schools
Street             City     State Zip                                       
8809 Flair Square  Toddside IL    37206          C-      B+       B-       A
9901 Austin Street Toddside IL    37206          D+       C        A      C+
905 Hogan Quarter  Franklin IL    37206          A-       A       D+       F

MultiIndex는 여러 level, 여러 계층을 저장할 수 있는 index 이다. 그리고 각 index label은 여러 구성 요소를 가진다.

## 7.2 MultiIndex DataFrames

pandas에서 누락된 열 header를 구별하기 위해 Unnamed: 0과 같이 나타낸다.

In [11]:
#원본 csv파일이라고 하겠다.
neighborhoods = pd.read_csv("neighborhoods.csv")
neighborhoods.head()

,Unnamed: 0,Unnamed: 1,Unnamed: 2,Culture,Culture.1,Services,Services.1
0,NaN,NaN,NaN,Restaurants,Museums,Police,Schools
1,State,City,Street,NaN,NaN,NaN,NaN
2,MO,Fisherborough,244 Tracy View,C+,F,D-,A+
3,SD,Port Curtisville,446 Cynthia Inlet,C-,B,B,D+
4,WV,Jimenezview,432 John Common,A,A+,F,B


In [12]:
#가장 왼쪽에 있는 3개의 열이 DataFrame의 index 역할을 한도록 한다.
#이를 위해 그에 해당하는 열의 index를 나타내는 숫자 list를 매개변수 index_col에 전달해야 한다.
#incdex_col에 여러 값이 있는 list가 전달되면 이에 대한 MultiIndex를 자동으로 생성한다.
neighborhoods = pd.read_csv(
    "neighborhoods.csv",
    index_col = [0, 1, 2]
)

neighborhoods.head()

,,,Culture,Culture.1,Services,Services.1
NaN,NaN,NaN,Restaurants,Museums,Police,Schools
State,City,Street,NaN,NaN,NaN,NaN
MO,Fisherborough,244 Tracy View,C+,F,D-,A+
SD,Port Curtisville,446 Cynthia Inlet,C-,B,B,D+
WV,Jimenezview,432 John Common,A,A+,F,B


In [13]:
#다음으로 DataFrame의 header에 사용할 dataset 행을 pandas에게 알려준다.
#위 원본 csv파일을 봤을 때 0행과 1행을 column header로 설정해야 하는데, 이를 위해 header에 [0,1]를 전달한다.
#그럼 열에 MultiIndex를 할당한다.
neighborhoods = pd.read_csv(
    "neighborhoods.csv",
    index_col = [0, 1, 2],     #0,1,2번째 열이 MultiIndex로써 사용되고
    header = [0, 1]            #0행과 1행이 각각 column header로 사용된다.
)

neighborhoods.head()

Culture         Services        
                                         Restaurants Museums   Police Schools
State City             Street                                                
MO    Fisherborough    244 Tracy View             C+       F       D-      A+
SD    Port Curtisville 446 Cynthia Inlet          C-       B        B      D+
WV    Jimenezview      432 John Common             A      A+        F       B
AK    Stevenshire      238 Andrew Rue             D-       A       A-      A-
ND    New Joshuaport   877 Walter Neck            D+      C-        B       B

In [14]:
neighborhoods.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 251 entries, ('MO', 'Fisherborough', '244 Tracy View') to ('NE', 'South Kennethmouth', '346 Wallace Pass')
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   (Culture, Restaurants)  251 non-null    object
 1   (Culture, Museums)      251 non-null    object
 2   (Services, Police)      251 non-null    object
 3   (Services, Schools)     251 non-null    object
dtypes: object(4)
memory usage: 27.1+ KB


    MultiIndex: 251 entries, ('MO', 'Fisherborough', '244 Tracy View') to ('NE', 'South Kennethmouth', '346 Wallace Pass')
    Data columns (total 4 columns):
         0   (Culture, Restaurants)  251 non-null    object
         1   (Culture, Museums)      251 non-null    object
         2   (Services, Police)      251 non-null    object
         3   (Services, Schools)     251 non-null    object

위와 같이 각 열의 이름을 (Culture, Restaurants)와 같이 2개의 요소를 가진 tuple로 출력한다.

마찬가지로 각 행의 label를 ('MO', 'Fisherborough', '244 Tracy View')과 같이 3개의 요소를 가진 tuple로 출력한다.

In [15]:
#index속성을 사용하여 행의 MultiIndex object(index label)에 접근할 수 있다.
print(type(neighborhoods.index))
neighborhoods.index

<class 'pandas.core.indexes.multi.MultiIndex'>


MultiIndex([('MO',      'Fisherborough',        '244 Tracy View'),
            ('SD',   'Port Curtisville',     '446 Cynthia Inlet'),
            ('WV',        'Jimenezview',       '432 John Common'),
            ('AK',        'Stevenshire',        '238 Andrew Rue'),
            ('ND',     'New Joshuaport',       '877 Walter Neck'),
            ('ID',         'Wellsville',   '696 Weber Stravenue'),
            ('TN',          'Jodiburgh',    '285 Justin Corners'),
            ('DC',   'Lake Christopher',   '607 Montoya Harbors'),
            ('OH',          'Port Mike',      '041 Michael Neck'),
            ('ND',         'Hardyburgh', '550 Gilmore Mountains'),
            ...
            ('AK',          'Scottstad',      '114 Jones Garden'),
            ('IA',    'Port Willieport',  '320 Jennifer Mission'),
            ('ME',         'Port Linda',        '692 Hill Glens'),
            ('KS',         'Kaylamouth',       '483 Freeman Via'),
            ('WA',     'Port Shawnfort',    '6

In [16]:
#columns속성을 사용하여 열의 MultiIndex object에 접근할 수 있다.
neighborhoods.columns

MultiIndex([( 'Culture', 'Restaurants'),
            ( 'Culture',     'Museums'),
            ('Services',      'Police'),
            ('Services',     'Schools')],
           )

In [17]:
#MultiIndex object의 names속성을 사용하여 index 이름 list에 접근할 수 있다.
neighborhoods.index.names

FrozenList(['State', 'City', 'Street'])

In [18]:
neighborhoods.head(3)

Culture         Services        
                                         Restaurants Museums   Police Schools
State City             Street                                                
MO    Fisherborough    244 Tracy View             C+       F       D-      A+
SD    Port Curtisville 446 Cynthia Inlet          C-       B        B      D+
WV    Jimenezview      432 John Common             A      A+        F       B

In [19]:
#MultiIndex object의 get_level_values()는 index object를 추출한다.
#이때 인자로는 level의 index위치나 level의 이름을 전달한다.

#다음은 "City"라는 level에 대한 각 행의 값을 추출한다.

# The two lines below are equivalent
neighborhoods.index.get_level_values(1)
neighborhoods.index.get_level_values("City")

Index(['Fisherborough', 'Port Curtisville', 'Jimenezview', 'Stevenshire',
       'New Joshuaport', 'Wellsville', 'Jodiburgh', 'Lake Christopher',
       'Port Mike', 'Hardyburgh',
       ...
       'Scottstad', 'Port Willieport', 'Port Linda', 'Kaylamouth',
       'Port Shawnfort', 'North Matthew', 'Chadton', 'Diazmouth', 'Laurentown',
       'South Kennethmouth'],
      dtype='object', name='City', length=251)

In [20]:
#CSV가 이름을 제공하지 않았기에 열의 MultiIndex level에는 이름이 없다.
#즉 (Culture, Services)열에 대한 이름과 (Restaurants, Museums, Police, Schools)열에 대한 이름이 없다.
neighborhoods.columns.names

FrozenList([None, None])

In [21]:
neighborhoods.columns.names = ["Category", "Subcategory"]
neighborhoods.columns.names

FrozenList(['Category', 'Subcategory'])

In [22]:
#이제는 (Culture, Services)열에 대한 이름과 (Restaurants, Museums, Police, Schools)열에 대한 이름이 존재함을 볼 수 있다.
neighborhoods.head(3)

Category                                     Culture         Services        
Subcategory                              Restaurants Museums   Police Schools
State City             Street                                                
MO    Fisherborough    244 Tracy View             C+       F       D-      A+
SD    Port Curtisville 446 Cynthia Inlet          C-       B        B      D+
WV    Jimenezview      432 John Common             A      A+        F       B

In [23]:
#columns.get_level_values()이므로 열의 MultiIndex에서 index를 검색할 수 있다.
#즉 Category의 index를 검색할 수 있다.

# The two lines below are equivalent
neighborhoods.columns.get_level_values(0)
neighborhoods.columns.get_level_values("Category")

Index(['Culture', 'Culture', 'Services', 'Services'], dtype='object', name='Category')

In [24]:
neighborhoods.head(1)

Category                               Culture         Services        
Subcategory                        Restaurants Museums   Police Schools
State City          Street                                             
MO    Fisherborough 244 Tracy View          C+       F       D-      A+

In [25]:
#nunique()는 각 열의 고유값 개수를 나타내는 Series를 반환한다.
print(type(neighborhoods.nunique()))
neighborhoods.nunique()

<class 'pandas.core.series.Series'>


Category  Subcategory
Culture   Restaurants    13
          Museums        13
Services  Police         13
          Schools        13
dtype: int64

In [26]:
#행이 MultiIndex로 이뤄진 Series에서 값에 접근할 때는 다음과 같이 접근하면 된다.
(neighborhoods.nunique())[("Culture","Restaurants")]

np.int64(13)

## 7.3 Sorting a MultiIndex

pandas는 순서가 뒤죽박죽인 컬렉션보다 정렬된 컬렉션에서 훨씬 빠르게 값을 찾을 수 있다.

In [27]:
#MultiIndex DataFrame에서 sort_index()를 호출할 때 pandas는 모든 level을 오름차순으로 정렬하고 밖에서부터 안쪽 방향으로 정렬을 진행한다.
#지금도 보면 State를 먼저 정렬하고, 정렬된 State를 기준으로 City를 정렬 또 이를 기준으로 Street를 정렬한다.
neighborhoods.sort_index()

Category                                      Culture         Services        
Subcategory                               Restaurants Museums   Police Schools
State City           Street                                                   
AK    Rowlandchester 386 Rebecca Cove              C-      A-       A+       C
      Scottstad      082 Leblanc Freeway            D      C-        D      B+
                     114 Jones Garden              D-      D-        D       D
      Stevenshire    238 Andrew Rue                D-       A       A-      A-
AL    Clarkland      430 Douglas Mission            A       F       C+      B+
...                                               ...     ...      ...     ...
WY    Lake Nicole    754 Weaver Turnpike            B      D-        B       D
                     933 Jennifer Burg              C      A+       A-       C
      Martintown     013 Bell Mills                C-       D       A-      B-
      Port Jason     624 Faulkner Orchard          A-       F       C+      C+
      Reneeshire     717 Patel Square               B      B+        D       A

[251 rows x 4 columns]

In [28]:
#ascending로 오름차순으로 정렬할지 내림차순으로 정렬할지 결정한다.
neighborhoods.sort_index(ascending = False).head()

Category                                   Culture         Services        
Subcategory                            Restaurants Museums   Police Schools
State City        Street                                                   
WY    Reneeshire  717 Patel Square               B      B+        D       A
      Port Jason  624 Faulkner Orchard          A-       F       C+      C+
      Martintown  013 Bell Mills                C-       D       A-      B-
      Lake Nicole 933 Jennifer Burg              C      A+       A-       C
                  754 Weaver Turnpike            B      D-        B       D

In [29]:
#각 level의 정렬 순서를 서로 다르게 지정할 수도 있다.
#State는 오름차순, City는 내림차순, Street는 오름차순으로 정렬하도록 설정하였다.
neighborhoods.sort_index(ascending = [True, False, True]).head()

Category                                     Culture         Services        
Subcategory                              Restaurants Museums   Police Schools
State City           Street                                                  
AK    Stevenshire    238 Andrew Rue               D-       A       A-      A-
      Scottstad      082 Leblanc Freeway           D      C-        D      B+
                     114 Jones Garden             D-      D-        D       D
      Rowlandchester 386 Rebecca Cove             C-      A-       A+       C
AL    Vegaside       191 Mindy Meadows            B+      A-       A+      D+

In [30]:
#MultiIndex level자체를 정렬할 수도 있다. 이때는 City의 값을 기준으로 행을 정렬하겠다.
#level매개변수에 level의 index위치나 이름을 전달한다.
#이때는 City level만 오름차순으로 정렬된다.

# The two lines below are equivalent
neighborhoods.sort_index(level = 1)
neighborhoods.sort_index(level = "City")

Category                                      Culture         Services        
Subcategory                               Restaurants Museums   Police Schools
State City          Street                                                    
AR    Allisonland   124 Diaz Brooks                C-      A+        F      C+
GA    Amyburgh      941 Brian Expressway            B       B       D-      C+
IA    Amyburgh      163 Heather Neck                F       D       A+      A-
ID    Andrewshire   952 Ellis Drive                C+      A-       C+       A
UT    Baileyfort    919 Stewart Hills              D+      C+        A       C
...                                               ...     ...      ...     ...
NC    West Scott    348 Jack Branch                A-      D-       A-       A
SD    West Scott    139 Hardy Vista                C+      A-       D+      B-
IN    Wilsonborough 066 Carr Road                  A+      C-        B       F
NC    Wilsonshire   871 Christopher Vista          B+       B       D+       F
NV    Wilsonshire   542 Jessica Stream              A      A+       C-      C+

[251 rows x 4 columns]

In [31]:
#다음은 City와 Street level의 값만 정렬하도록 한다. City정렬하고 이를 기준으로 Street를 정렬하도록 한다.

# The two lines below are equivalent
neighborhoods.sort_index(level = [1, 2]).head()
neighborhoods.sort_index(level = ["City", "Street"]).head()

Category                                   Culture         Services        
Subcategory                            Restaurants Museums   Police Schools
State City        Street                                                   
AR    Allisonland 124 Diaz Brooks               C-      A+        F      C+
IA    Amyburgh    163 Heather Neck               F       D       A+      A-
GA    Amyburgh    941 Brian Expressway           B       B       D-      C+
ID    Andrewshire 952 Ellis Drive               C+      A-       C+       A
VT    Baileyfort  831 Norma Cove                 B      D+       A+      D+

In [32]:
#City는 오름차순으로 Street는 내림차순으로 정렬하도록 한다(2,3번째 행을 보면 어떻게 적용되었는지 알 수 있다.).
neighborhoods.sort_index(
    level = ["City", "Street"], ascending = [True, False]
).head()

Category                                   Culture         Services        
Subcategory                            Restaurants Museums   Police Schools
State City        Street                                                   
AR    Allisonland 124 Diaz Brooks               C-      A+        F      C+
GA    Amyburgh    941 Brian Expressway           B       B       D-      C+
IA    Amyburgh    163 Heather Neck               F       D       A+      A-
ID    Andrewshire 952 Ellis Drive               C+      A-       C+       A
UT    Baileyfort  919 Stewart Hills             D+      C+        A       C

In [33]:
#정렬할려는 level의 수와 ascending에 전달되는 요소의 개수가 다를 때는 오류가 발생한다.
neighborhoods.sort_index(
    level = ["City", "Street"], ascending = [True,True, False]
).head()

ValueError: level must have same length as ascending

In [ ]:
#축을 column으로 설정하면 열의 MultiIndex도 정렬할 수 있다.
#다은은 Category level을 먼저 정렬하고 Subcategory level을 두 번째로 정렬한다.

# The two lines below are equivalent
neighborhoods.sort_index(axis = 1).head(3)
neighborhoods.sort_index(axis = "columns").head(3)

In [34]:
#Subcategory level 값만을 내림차순으로 정렬한다.
#그럼 Restaurants	Museums	Police	Schools 이 Schools	Restaurants	Police	Museums 로 정렬되는데
#Category그룹이 번갈아 나타나기에 다음과 같이 출력이 된다.

neighborhoods.sort_index(
    axis = 1, level = "Subcategory", ascending = False
).head(3)

,,Category,Services,Culture,Services,Culture
,,Subcategory,Schools,Restaurants,Police,Museums
State,City,Street,,,,
MO,Fisherborough,244 Tracy View,A+,C+,D-,F
SD,Port Curtisville,446 Cynthia Inlet,D+,C-,B,B
WV,Jimenezview,432 John Common,B,A,F,A+


In [35]:
neighborhoods = neighborhoods.sort_index(ascending = True)

In [36]:
neighborhoods.head(3)

Category                                     Culture         Services        
Subcategory                              Restaurants Museums   Police Schools
State City           Street                                                  
AK    Rowlandchester 386 Rebecca Cove             C-      A-       A+       C
      Scottstad      082 Leblanc Freeway           D      C-        D      B+
                     114 Jones Garden             D-      D-        D       D

## 7.4 Selecting with a MultiIndex

In [38]:
data = [
    [1, 2], 
    [3, 4]
]

df = pd.DataFrame(
    data = data,
    index = ["A", "B"],
    columns = ["X", "Y"]
)

df

,X,Y
A,1,2
B,3,4


In [39]:
#열ㅇ르 Series로 추출할 수 있다.
df["X"]

A    1
B    3
Name: X, dtype: int64

### 7.4.1 Extracting One or More Columns

In [41]:
neighborhoods.head(0)

Empty DataFrame
Columns: [(Culture, Restaurants), (Culture, Museums), (Services, Police), (Services, Schools)]
Index: []

In [44]:
#Services열에 대한 DataFrame를 추출한다.
#Services는 Police와 Schools를 하위 범주로 갖고 있기에 다음과 같이 나타난다.
print(type(neighborhoods["Services"]))
neighborhoods["Services"]

<class 'pandas.core.frame.DataFrame'>


Subcategory                               Police Schools
State City           Street                             
AK    Rowlandchester 386 Rebecca Cove         A+       C
      Scottstad      082 Leblanc Freeway       D      B+
                     114 Jones Garden          D       D
      Stevenshire    238 Andrew Rue           A-      A-
AL    Clarkland      430 Douglas Mission      C+      B+
...                                          ...     ...
WY    Lake Nicole    754 Weaver Turnpike       B       D
                     933 Jennifer Burg        A-       C
      Martintown     013 Bell Mills           A-      B-
      Port Jason     624 Faulkner Orchard     C+      C+
      Reneeshire     717 Patel Square          D       A

[251 rows x 2 columns]

**NOTE**: I've commented out the code below so that the Notebook can run without raising an error.

In [45]:
neighborhoods["Schools"]

KeyError: 'Schools'

In [47]:
#School은 Services의 하위 열이므로 항상 ("Services","Schools")과 같이 호출해야 한다.
print(type(neighborhoods[("Services", "Schools")]))
neighborhoods[("Services", "Schools")]

<class 'pandas.core.series.Series'>


State  City            Street              
AK     Rowlandchester  386 Rebecca Cove         C
       Scottstad       082 Leblanc Freeway     B+
                       114 Jones Garden         D
       Stevenshire     238 Andrew Rue          A-
AL     Clarkland       430 Douglas Mission     B+
                                               ..
WY     Lake Nicole     754 Weaver Turnpike      D
                       933 Jennifer Burg        C
       Martintown      013 Bell Mills          B-
       Port Jason      624 Faulkner Orchard    C+
       Reneeshire      717 Patel Square         A
Name: (Services, Schools), Length: 251, dtype: object

In [48]:
#여러 개의 열을 추출하기 위해서는 tuple의 list를 대괄호 안에 넣으면 된다.
#list에 있는 tuple의 순서는 결과 DataFrame의 열 순서를 결정한다.
neighborhoods[[("Services", "Schools"), ("Culture", "Museums")]]

Category                                  Services Culture
Subcategory                                Schools Museums
State City           Street                               
AK    Rowlandchester 386 Rebecca Cove            C      A-
      Scottstad      082 Leblanc Freeway        B+      C-
                     114 Jones Garden            D      D-
      Stevenshire    238 Andrew Rue             A-       A
AL    Clarkland      430 Douglas Mission        B+       F
...                                            ...     ...
WY    Lake Nicole    754 Weaver Turnpike         D      D-
                     933 Jennifer Burg           C      A+
      Martintown     013 Bell Mills             B-       D
      Port Jason     624 Faulkner Orchard       C+       F
      Reneeshire     717 Patel Square            A      B+

[251 rows x 2 columns]

In [50]:
#위와 같은 결과를 나타낸다.
columns = [
    ("Services", "Schools"),
    ("Culture", "Museums")
]

neighborhoods[columns]

Category                                  Services Culture
Subcategory                                Schools Museums
State City           Street                               
AK    Rowlandchester 386 Rebecca Cove            C      A-
      Scottstad      082 Leblanc Freeway        B+      C-
                     114 Jones Garden            D      D-
      Stevenshire    238 Andrew Rue             A-       A
AL    Clarkland      430 Douglas Mission        B+       F
...                                            ...     ...
WY    Lake Nicole    754 Weaver Turnpike         D      D-
                     933 Jennifer Burg           C      A+
      Martintown     013 Bell Mills             B-       D
      Port Jason     624 Faulkner Orchard       C+       F
      Reneeshire     717 Patel Square            A      B+

[251 rows x 2 columns]

### 7.4.2 Extracting One or More Rows with loc

In [51]:
df

,X,Y
A,1,2
B,3,4


In [52]:
df.loc["A"]

X    1
Y    2
Name: A, dtype: int64

In [53]:
df.iloc[1]

X    3
Y    4
Name: B, dtype: int64

In [57]:
#neighborhoods DataFrame의 MultiIndex에는 State, City와 Address라는 세 가지 level이 있다.
#특정 행에 대한 level값을 알고 있다면 이를 통해(이를 index label로 사용하여) 그 행의 값을 찾을 수 있다.

#다음은 State, City와 Address라는 세 가지 level값을 모두 알고 있을 때이다.
neighborhoods.loc[("TX", "Kingchester", "534 Gordon Falls")]

Category  Subcategory
Culture   Restaurants     C
          Museums        D+
Services  Police          B
          Schools         B
Name: (TX, Kingchester, 534 Gordon Falls), dtype: object

In [56]:
#State level값만 알고 있을 때
neighborhoods.loc["CA"]

Category                              Culture         Services        
Subcategory                       Restaurants Museums   Police Schools
City           Street                                                 
Dustinmouth    793 Cynthia Square          A-      A+       C-       A
North Jennifer 303 Alisha Road             D-      C+       C+      A+
Ryanfort       934 David Run                F      B+        F      D-

In [58]:
#State level값과 City level값을 알고 있을 때
neighborhoods.loc["CA", "Dustinmouth"]

Category               Culture         Services        
Subcategory        Restaurants Museums   Police Schools
Street                                                 
793 Cynthia Square          A-      A+       C-       A

In [60]:
#loc의 두 번째 인수로 추출할 열을 지정할 수도 있다.
#State값이 CA이고 열 MultiIndex에서 Category값이 Culture인 행을 추출한다.
neighborhoods.loc["CA", "Culture"]

,Subcategory,Restaurants,Museums
City,Street,,
Dustinmouth,793 Cynthia Square,A-,A+
North Jennifer,303 Alisha Road,D-,C+
Ryanfort,934 David Run,F,B+


In [61]:
#헷갈리지 않도록 첫 번째 인수로는 행 index label로 사용하고
#두 번째 인수로는 열 index label로 사용한다.
neighborhoods.loc[("CA", "Dustinmouth")]

Category               Culture         Services        
Subcategory        Restaurants Museums   Police Schools
Street                                                 
793 Cynthia Square          A-      A+       C-       A

In [64]:
#행이나 열의 index label을 여러 개 지정해야 할 때에는 tuple을 사용한다.
neighborhoods.loc[("CA", "Dustinmouth"), ("Services",)]

Subcategory,Police,Schools
Street,,
793 Cynthia Square,C-,A


In [65]:
neighborhoods.loc[("CA", "Dustinmouth"), ("Services", "Schools")]

Street
793 Cynthia Square    A
Name: (Services, Schools), dtype: object

In [66]:
#State 값이 NE와 NH사이인 모든 연속된 행을 가져온다.
neighborhoods["NE":"NH"]

Empty DataFrame
Columns: []
Index: [(AK, Rowlandchester, 386 Rebecca Cove), (AK, Scottstad, 082 Leblanc Freeway), (AK, Scottstad, 114 Jones Garden), (AK, Stevenshire, 238 Andrew Rue), (AL, Clarkland, 430 Douglas Mission), (AL, East Bobbyside, 294 Mitchell Village), (AL, East Connie, 943 John Gardens), (AL, Kathrynstad, 356 Dodson Plains), (AL, Lonniestad, 643 Allen Estates), (AL, Port Jenniferfurt, 539 Horton Fork), (AL, Samuelville, 989 Dean Plains), (AL, Vegaside, 191 Mindy Meadows), (AR, Allisonland, 124 Diaz Brooks), (AR, New Larryshire, 372 Vazquez Isle), (AR, Russoshire, 499 Hinton Forks), (AR, Simmonsfurt, 154 Alexander Track), (AZ, Lake Davidshire, 943 Melissa Haven), (AZ, Tiffanyberg, 344 Ronald Groves), (AZ, West Richard, 847 Nelson Cape), (CA, Dustinmouth, 793 Cynthia Square), (CA, North Jennifer, 303 Alisha Road), (CA, Ryanfort, 934 David Run), (CO, Lake Robert, 049 Jordan Lane), (CO, Port Jeanette, 219 Henderson Glen), (CO, South Eugene, 654 Fisher Summit), (CT, East Jessicaland, 208 Todd Knolls), (CT, New Adrianhaven, 048 Brian Cove), (CT, Port Mike, 410 Keith Lodge), (CT, Sethstad, 139 Bailey Grove), (DC, East Jessica, 149 Norman Crossing), (DC, East Lisaview, 910 Sandy Ramp), (DC, Lake Christopher, 607 Montoya Harbors), (DC, New Janice, 477 Flores Cliff), (DC, Richardland, 419 Michael Squares), (DC, South Nathan, 631 Lynch Hill), (DE, Dustinmouth, 719 Dixon Ways), (DE, Lake Davidmouth, 051 Destiny Mission), (DE, Moonborough, 518 Christopher Springs), (DE, New Christopher, 822 Connie Dale), (DE, New Janice, 417 Simmons Spur), (DE, Port Michael, 328 Brian Cliff), (DE, Port Michael, 800 Christopher Haven), (DE, South Kayla, 731 Atkinson Knolls), (DE, South Tara, 047 Martinez Keys), (FL, Bryanchester, 481 Julie Meadows), (FL, Jacksonfort, 369 Victoria Road), (FL, Kingchester, 467 Lisa Mission), (FL, Port Melissaburgh, 393 Jennifer Parkway), (FL, West Miguelstad, 570 James Unions), (GA, Amyburgh, 941 Brian Expressway), (GA, New Carmenland, 943 James Lake), (GA, Spencechester, 570 Davis Isle), (GA, Tinaborough, 946 Lindsay Port), (GA, Trevorville, 283 Kelli Forest), (HI, Kathleenport, 507 White Vista), (HI, Lauratown, 595 Moore Corner), (HI, New Caroline, 302 Paul Spring), (HI, New Jennifer, 931 Pace Lights), (HI, Port Ronnietown, 323 William Club), (HI, Port Ronnietown, 710 Timothy Mountain), (HI, South Lauren, 073 Amanda Well), (HI, Timothystad, 393 Jasmine Hollow), (IA, Amyburgh, 163 Heather Neck), (IA, Chadburgh, 877 Smith Haven), (IA, East Brittany, 690 Gonzales Tunnel), (IA, East Christopher, 955 Hoffman Camp), (IA, Port Willieport, 320 Jennifer Mission), (ID, Andrewshire, 952 Ellis Drive), (ID, Patricktown, 215 Kathleen Flat), (ID, Spencechester, 523 Ali Stream), (ID, Thompsonview, 028 Ford Brook), (ID, Wellsville, 696 Weber Stravenue), (ID, West Anna, 544 Patel Locks), (ID, West Christophertown, 168 Brown Stravenue), (IN, East Deborahberg, 829 Melissa Mountain), (IN, Fisherborough, 975 Bell Fork), (IN, New Carmenland, 567 Ricky Trafficway), (IN, Simmonsfurt, 547 Miller Isle), (IN, Smithside, 347 Nicholas Shores), (IN, Wilsonborough, 066 Carr Road), (KS, Kaylamouth, 483 Freeman Via), (KS, Port Lindaberg, 008 Mitchell Highway), (KS, Shawnchester, 103 David Crest), (KS, South Karenberg, 141 Eric Lock), (KS, Timothystad, 194 Miller Grove), (KY, Bowmantown, 698 Lee Meadows), (KY, Hammondside, 751 Patton Squares), (LA, Danielmouth, 084 Jamie Tunnel), (LA, Jimenezview, 063 Cox Park), (LA, Port Michellemouth, 155 Ethan Island), (MA, Danielmouth, 539 Curry Station), (MA, Gonzalesbury, 770 Little Shoals), (MA, Stevenshire, 660 Anthony Path), (MD, Richardland, 695 Amy Ports), (MD, Sanchezmouth, 003 Daniel Via), (MD, West Miguelstad, 751 Walker Meadows), (ME, Kathleenport, 520 Ellis Meadows), (ME, Lake Ambertown, 654 Margaret Lake), (ME, Lake Timothy, 746 Jeffrey Bypass), (ME, Morrisonburgh, 335 Scott Manors), ...]

[251 rows x 0 columns]

In [67]:
#State값이 NE이고 City값이 Shawnchester인 행부터
#State값이 NH이고 City값이 North Latoya인 행까지 모든 행을 추출한다.
neighborhoods.loc[("NE", "Shawnchester"):("NH", "North Latoya")]

Category                                          Culture         Services  \
Subcategory                                   Restaurants Museums   Police   
State City               Street                                              
NE    Shawnchester       802 Cook Cliff                D-      D+        D   
      South Kennethmouth 346 Wallace Pass              C-      B-        A   
      South Nathan       821 Jake Fork                 C+       D       D+   
NH    Courtneyfort       697 Spencer Isle              A+      A+       C+   
      East Deborahberg   271 Ryan Mount                 B       C       D+   
      Ingramton          430 Calvin Underpass          C+      D+        C   
      North Latoya       603 Clark Mount               D-      A-       B+   

Category                                               
Subcategory                                   Schools  
State City               Street                        
NE    Shawnchester       802 Cook Cliff             A  
      South Kennethmouth 346 Wallace Pass          A-  
      South Nathan       821 Jake Fork              A  
NH    Courtneyfort       697 Spencer Isle          A+  
      East Deborahberg   271 Ryan Mount            B-  
      Ingramton          430 Calvin Underpass      C-  
      North Latoya       603 Clark Mount           B-

In [68]:
start = ("NE", "Shawnchester")
end   = ("NH", "North Latoya")
neighborhoods.loc[start:end]

Category                                          Culture         Services  \
Subcategory                                   Restaurants Museums   Police   
State City               Street                                              
NE    Shawnchester       802 Cook Cliff                D-      D+        D   
      South Kennethmouth 346 Wallace Pass              C-      B-        A   
      South Nathan       821 Jake Fork                 C+       D       D+   
NH    Courtneyfort       697 Spencer Isle              A+      A+       C+   
      East Deborahberg   271 Ryan Mount                 B       C       D+   
      Ingramton          430 Calvin Underpass          C+      D+        C   
      North Latoya       603 Clark Mount               D-      A-       B+   

Category                                               
Subcategory                                   Schools  
State City               Street                        
NE    Shawnchester       802 Cook Cliff             A  
      South Kennethmouth 346 Wallace Pass          A-  
      South Nathan       821 Jake Fork              A  
NH    Courtneyfort       697 Spencer Isle          A+  
      East Deborahberg   271 Ryan Mount            B-  
      Ingramton          430 Calvin Underpass      C-  
      North Latoya       603 Clark Mount           B-

In [70]:
#State값이 NE이고 City값이 Shawnchester인 행부터
#State값이 NH이고 City값의 끝까지 모든 행을 추출한다.

neighborhoods.loc[("NE", "Shawnchester"):("NH")]

Category                                          Culture         Services  \
Subcategory                                   Restaurants Museums   Police   
State City               Street                                              
NE    Shawnchester       802 Cook Cliff                D-      D+        D   
      South Kennethmouth 346 Wallace Pass              C-      B-        A   
      South Nathan       821 Jake Fork                 C+       D       D+   
NH    Courtneyfort       697 Spencer Isle              A+      A+       C+   
      East Deborahberg   271 Ryan Mount                 B       C       D+   
      Ingramton          430 Calvin Underpass          C+      D+        C   
      North Latoya       603 Clark Mount               D-      A-       B+   
      South Tara         559 Michael Glens             C-      C-        F   

Category                                               
Subcategory                                   Schools  
State City               Street                        
NE    Shawnchester       802 Cook Cliff             A  
      South Kennethmouth 346 Wallace Pass          A-  
      South Nathan       821 Jake Fork              A  
NH    Courtneyfort       697 Spencer Isle          A+  
      East Deborahberg   271 Ryan Mount            B-  
      Ingramton          430 Calvin Underpass      C-  
      North Latoya       603 Clark Mount           B-  
      South Tara         559 Michael Glens          B

### 7.4.3 Extracting One or More Rows with iloc

In [71]:
#index위치가 25인 행을 추출
neighborhoods.iloc[25]

Category  Subcategory
Culture   Restaurants    A+
          Museums         A
Services  Police         A+
          Schools        C+
Name: (CT, East Jessicaland, 208 Todd Knolls), dtype: object

In [72]:
#index위치가 25인 행과 index위치가 2인 열을 추출
neighborhoods.iloc[25, 2]

'A+'

In [76]:
#index위치가 25인 행과 30인 행을 추출
neighborhoods.iloc[[25, 30]]

Category                                   Culture         Services        
Subcategory                            Restaurants Museums   Police Schools
State City             Street                                              
CT    East Jessicaland 208 Todd Knolls          A+       A       A+      C+
DC    East Lisaview    910 Sandy Ramp           A-      A+        B       B

In [77]:
#25행부터 29행까지 추출
neighborhoods.iloc[25:30]

Category                                       Culture         Services  \
Subcategory                                Restaurants Museums   Police   
State City             Street                                             
CT    East Jessicaland 208 Todd Knolls              A+       A       A+   
      New Adrianhaven  048 Brian Cove               A-      C+       A+   
      Port Mike        410 Keith Lodge              D-       A       B+   
      Sethstad         139 Bailey Grove              C      C-       C+   
DC    East Jessica     149 Norman Crossing          A-      C-       C+   

Category                                            
Subcategory                                Schools  
State City             Street                       
CT    East Jessicaland 208 Todd Knolls          C+  
      New Adrianhaven  048 Brian Cove           D-  
      Port Mike        410 Keith Lodge           D  
      Sethstad         139 Bailey Grove         A+  
DC    East Jessica     149 Norman Crossing      A-

In [78]:
#25행부터 29행까지 이때 1에서 2열까지의 값을 가져온다.
neighborhoods.iloc[25:30, 1:3]

Category                                   Culture Services
Subcategory                                Museums   Police
State City             Street                              
CT    East Jessicaland 208 Todd Knolls           A       A+
      New Adrianhaven  048 Brian Cove           C+       A+
      Port Mike        410 Keith Lodge           A       B+
      Sethstad         139 Bailey Grove         C-       C+
DC    East Jessica     149 Norman Crossing      C-       C+

In [79]:
neighborhoods.iloc[-4:, -2:]

Category                               Services        
Subcategory                              Police Schools
State City        Street                               
WY    Lake Nicole 933 Jennifer Burg          A-       C
      Martintown  013 Bell Mills             A-      B-
      Port Jason  624 Faulkner Orchard       C+      C+
      Reneeshire  717 Patel Square            D       A

## 7.5 Cross Sections

In [80]:
#xs()에 하나의 MultiIndex level에 대한 값을 넘기면 그에 대한 행을 추출할 수 있다.
#찾고자 하는 값을 key매개변수에 전달하고 level에는 index level의 이름이나 숫자 위치를 전달한다.

# The two lines below are equivalent
neighborhoods.xs(key = "Lake Nicole", level = 1)
neighborhoods.xs(key = "Lake Nicole", level = "City")

Category                      Culture         Services        
Subcategory               Restaurants Museums   Police Schools
State Street                                                  
OR    650 Angela Track              D      C-        D       F
WY    754 Weaver Turnpike           B      D-        B       D
      933 Jennifer Burg             C      A+       A-       C

In [81]:
#열의 MultiIndex에 있는 Subcategory level에서 Museums key가 있는 열을 선택한다.

neighborhoods.xs(
    axis = "columns", key = "Museums", level = "Subcategory"
).head()

Category                                 Culture
State City           Street                     
AK    Rowlandchester 386 Rebecca Cove         A-
      Scottstad      082 Leblanc Freeway      C-
                     114 Jones Garden         D-
      Stevenshire    238 Andrew Rue            A
AL    Clarkland      430 Douglas Mission       F

In [83]:
neighborhoods.xs(
    axis="columns",key="Services",level="Category").head()

Subcategory                              Police Schools
State City           Street                            
AK    Rowlandchester 386 Rebecca Cove        A+       C
      Scottstad      082 Leblanc Freeway      D      B+
                     114 Jones Garden         D       D
      Stevenshire    238 Andrew Rue          A-      A-
AL    Clarkland      430 Douglas Mission     C+      B+

In [85]:
#State값이 AK이고 Street값이 238 Andrew Rue인 행을 찾는다.

# The two lines below are equivalent
neighborhoods.xs(
    key = ("AK", "238 Andrew Rue"), level = ["State", "Street"]
)

neighborhoods.xs(
    key = ("AK", "238 Andrew Rue"), level = [0, 2]
)

Category        Culture         Services        
Subcategory Restaurants Museums   Police Schools
City                                            
Stevenshire          D-       A       A-      A-

## 7.6. Manipulating the Index

### 7.6.1 Resetting the Index

In [86]:
neighborhoods.head()

Category                                     Culture         Services        
Subcategory                              Restaurants Museums   Police Schools
State City           Street                                                  
AK    Rowlandchester 386 Rebecca Cove             C-      A-       A+       C
      Scottstad      082 Leblanc Freeway           D      C-        D      B+
                     114 Jones Garden             D-      D-        D       D
      Stevenshire    238 Andrew Rue               D-       A       A-      A-
AL    Clarkland      430 Douglas Mission           A       F       C+      B+

In [87]:
#reorder_levels()은 MultiIndex level을 지정된 순서로 정렬한다.
#이때 순서는 order매개변수에 list형태로 전달하면 된다.
new_order = ["City", "State", "Street"]
neighborhoods.reorder_levels(order = new_order).head()

Category                                     Culture         Services        
Subcategory                              Restaurants Museums   Police Schools
City           State Street                                                  
Rowlandchester AK    386 Rebecca Cove             C-      A-       A+       C
Scottstad      AK    082 Leblanc Freeway           D      C-        D      B+
                     114 Jones Garden             D-      D-        D       D
Stevenshire    AK    238 Andrew Rue               D-       A       A-      A-
Clarkland      AL    430 Douglas Mission           A       F       C+      B+

In [88]:
#위와 같은 결과를 출력한다.
neighborhoods.reorder_levels(order = [1, 0, 2]).head()

Category                                     Culture         Services        
Subcategory                              Restaurants Museums   Police Schools
City           State Street                                                  
Rowlandchester AK    386 Rebecca Cove             C-      A-       A+       C
Scottstad      AK    082 Leblanc Freeway           D      C-        D      B+
                     114 Jones Garden             D-      D-        D       D
Stevenshire    AK    238 Andrew Rue               D-       A       A-      A-
Clarkland      AL    430 Douglas Mission           A       F       C+      B+

In [90]:
#reset_index()는 이전 MultiIndex level을 열로 통합하는 새 DataFrame을 반환한다.

#결과를 보면 각 행은 index label이 아닌 index번호로 구분되는 것을 알 수 있다.
neighborhoods.reset_index().tail()

Category    State         City                Street     Culture          \
Subcategory                                          Restaurants Museums   
246            WY  Lake Nicole   754 Weaver Turnpike           B      D-   
247            WY  Lake Nicole     933 Jennifer Burg           C      A+   
248            WY   Martintown        013 Bell Mills          C-       D   
249            WY   Port Jason  624 Faulkner Orchard          A-       F   
250            WY   Reneeshire      717 Patel Square           B      B+   

Category    Services          
Subcategory   Police Schools  
246                B       D  
247               A-       C  
248               A-      B-  
249               C+      C+  
250                D       A

In [92]:
#State, City, Street 3개의 열을 다른 MultiIndex level에 추가할 수도 있다.
#원하는 level의 index 위치 또는 이름을 reset_index()의 col_level매개변수로 전달한다.

# The two lines below are equivalent
neighborhoods.reset_index(col_level = 1).tail()
neighborhoods.reset_index(col_level = "Subcategory").tail() 

Category                                                 Culture          \
Subcategory State         City                Street Restaurants Museums   
246            WY  Lake Nicole   754 Weaver Turnpike           B      D-   
247            WY  Lake Nicole     933 Jennifer Burg           C      A+   
248            WY   Martintown        013 Bell Mills          C-       D   
249            WY   Port Jason  624 Faulkner Orchard          A-       F   
250            WY   Reneeshire      717 Patel Square           B      B+   

Category    Services          
Subcategory   Police Schools  
246                B       D  
247               A-       C  
248               A-      B-  
249               C+      C+  
250                D       A

In [93]:
#State, City, Street의 상위 level Category에 빈 문자열을 할당하게 된다.
#이때 col_fill매개변수에 인수를 전달하면 빈 문자열을 원하는 값으로 바꿀 수 있다.
neighborhoods.reset_index(
    col_fill = "Address", col_level = "Subcategory"
).tail()

Category    Address                                        Culture          \
Subcategory   State         City                Street Restaurants Museums   
246              WY  Lake Nicole   754 Weaver Turnpike           B      D-   
247              WY  Lake Nicole     933 Jennifer Burg           C      A+   
248              WY   Martintown        013 Bell Mills          C-       D   
249              WY   Port Jason  624 Faulkner Orchard          A-       F   
250              WY   Reneeshire      717 Patel Square           B      B+   

Category    Services          
Subcategory   Police Schools  
246                B       D  
247               A-       C  
248               A-      B-  
249               C+      C+  
250                D       A

In [94]:
#level매개변수에 이름을 전달하여 그 이름에 대한 index level만 열로 옮길 수 있다.
neighborhoods.reset_index(level = "Street").tail()

Category                         Street     Culture         Services        
Subcategory                             Restaurants Museums   Police Schools
State City                                                                  
WY    Lake Nicole   754 Weaver Turnpike           B      D-        B       D
      Lake Nicole     933 Jennifer Burg           C      A+       A-       C
      Martintown         013 Bell Mills          C-       D       A-      B-
      Port Jason   624 Faulkner Orchard          A-       F       C+      C+
      Reneeshire       717 Patel Square           B      B+        D       A

In [72]:
neighborhoods.reset_index(level = ["Street", "City"]).tail()

Category            City                Street     Culture         Services  \
Subcategory                                    Restaurants Museums   Police   
State                                                                         
WY           Lake Nicole   754 Weaver Turnpike           B      D-        B   
WY           Lake Nicole     933 Jennifer Burg           C      A+       A-   
WY            Martintown        013 Bell Mills          C-       D       A-   
WY            Port Jason  624 Faulkner Orchard          A-       F       C+   
WY            Reneeshire      717 Patel Square           B      B+        D   

Category             
Subcategory Schools  
State                
WY                D  
WY                C  
WY               B-  
WY               C+  
WY                A

In [95]:
#reset_index()의 drop 매개변수 값을 True로 전달하면 pandas는 지정된 level을 열에 추가하는 대신 삭제한다.
neighborhoods.reset_index(level = "Street", drop = True).tail()

Category              Culture         Services        
Subcategory       Restaurants Museums   Police Schools
State City                                            
WY    Lake Nicole           B      D-        B       D
      Lake Nicole           C      A+       A-       C
      Martintown           C-       D       A-      B-
      Port Jason           A-       F       C+      C+
      Reneeshire            B      B+        D       A

In [96]:
neighborhoods = neighborhoods.reset_index()

In [97]:
neighborhoods

Category    State            City                Street     Culture          \
Subcategory                                             Restaurants Museums   
0              AK  Rowlandchester      386 Rebecca Cove          C-      A-   
1              AK       Scottstad   082 Leblanc Freeway           D      C-   
2              AK       Scottstad      114 Jones Garden          D-      D-   
3              AK     Stevenshire        238 Andrew Rue          D-       A   
4              AL       Clarkland   430 Douglas Mission           A       F   
..            ...             ...                   ...         ...     ...   
246            WY     Lake Nicole   754 Weaver Turnpike           B      D-   
247            WY     Lake Nicole     933 Jennifer Burg           C      A+   
248            WY      Martintown        013 Bell Mills          C-       D   
249            WY      Port Jason  624 Faulkner Orchard          A-       F   
250            WY      Reneeshire      717 Patel Square           B      B+   

Category    Services          
Subcategory   Police Schools  
0                 A+       C  
1                  D      B+  
2                  D       D  
3                 A-      A-  
4                 C+      B+  
..               ...     ...  
246                B       D  
247               A-       C  
248               A-      B-  
249               C+      C+  
250                D       A  

[251 rows x 7 columns]

### 7.6.2 Setting the Index

In [98]:
neighborhoods.head(3)

Category    State            City               Street     Culture          \
Subcategory                                            Restaurants Museums   
0              AK  Rowlandchester     386 Rebecca Cove          C-      A-   
1              AK       Scottstad  082 Leblanc Freeway           D      C-   
2              AK       Scottstad     114 Jones Garden          D-      D-   

Category    Services          
Subcategory   Police Schools  
0                 A+       C  
1                  D      B+  
2                  D       D

In [99]:
#set_index()는 하나 이상의 DataFrame열을 새로운 index로 설정한다.
#keys매개변수에 사용할 열을 전달한다.
neighborhoods.set_index(keys = "City").head()

Category       State               Street     Culture         Services        
Subcategory                               Restaurants Museums   Police Schools
City                                                                          
Rowlandchester    AK     386 Rebecca Cove          C-      A-       A+       C
Scottstad         AK  082 Leblanc Freeway           D      C-        D      B+
Scottstad         AK     114 Jones Garden          D-      D-        D       D
Stevenshire       AK       238 Andrew Rue          D-       A       A-      A-
Clarkland         AL  430 Douglas Mission           A       F       C+      B+

In [103]:
#MultiIndex level에서 대상으로 지정할 값을 담은 tuple을 keys 매개변수로 넘긴다.

#MultiIndex ("Culture", "Museums")를 index label로 지정하였다.
neighborhoods.set_index(keys = ("Culture", "Museums")).head()

Category           State            City               Street     Culture  \
Subcategory                                                   Restaurants   
(Culture, Museums)                                                          
A-                    AK  Rowlandchester     386 Rebecca Cove          C-   
C-                    AK       Scottstad  082 Leblanc Freeway           D   
D-                    AK       Scottstad     114 Jones Garden          D-   
A                     AK     Stevenshire       238 Andrew Rue          D-   
F                     AL       Clarkland  430 Douglas Mission           A   

Category           Services          
Subcategory          Police Schools  
(Culture, Museums)                   
A-                       A+       C  
C-                        D      B+  
D-                        D       D  
A                        A-      A-  
F                        C+      B+

In [104]:
#행 축에 MultiIndex를 생성하기 위해서는 여러 열이 있는 list를 keys매개변수로 전달한다.
neighborhoods.set_index(keys = ["State", "City"]).head()

Category                           Street     Culture         Services        
Subcategory                               Restaurants Museums   Police Schools
State City                                                                    
AK    Rowlandchester     386 Rebecca Cove          C-      A-       A+       C
      Scottstad       082 Leblanc Freeway           D      C-        D      B+
      Scottstad          114 Jones Garden          D-      D-        D       D
      Stevenshire          238 Andrew Rue          D-       A       A-      A-
AL    Clarkland       430 Douglas Mission           A       F       C+      B+

## 7.7 Coding Challenge

### 7.7.1 Problems

In [79]:
investments = pd.read_csv("investments.csv")
investments.head()

,Name,Market,Status,State,Funding Rounds
0,#waywire,News,Acquired,NY,1
1,&TV Communications,Games,Operating,CA,2
2,-R- Ranch and Mine,Tourism,Operating,TX,2
3,004 Technologies,Software,Operating,IL,1
4,1-4 All,Software,Operating,NC,1


In [80]:
investments.nunique()

Name              27763
Market              693
Status                3
State                61
Funding Rounds       16
dtype: int64

In [81]:
investments = investments.set_index(
    keys = ["Status", "Funding Rounds", "State"]
).sort_index()

In [82]:
investments.head()

Name               Market
Status   Funding Rounds State                                               
Acquired 1              AB               Hallpass Media                Games
                        AL                    EnteGreat  Enterprise Software
                        AL     Onward Behavioral Health        Biotechnology
                        AL                      Proxsys        Biotechnology
                        AZ                  Envox Group     Public Relations

### 7.7.2 Solutions

In [83]:
investments.loc[("Closed",)].head()

Name                Market
Funding Rounds State                                                   
1              AB     Cardinal Media Technologies  Social Network Media
               AB                Easy Bill Online              Tracking
               AB                   Globel Direct      Public Relations
               AB               Ph03nix New Media                 Games
               AL                           Naubo                  News

In [84]:
investments.loc[("Acquired", 10)]

,Name,Market
State,,
NY,Genesis Networks,Web Hosting
TX,ACTIVE Network,Software


In [85]:
investments.loc[("Operating", 6, "NJ")]

Name             Market
Status    Funding Rounds State                                       
Operating 6              NJ     Agile Therapeutics      Biotechnology
                         NJ               Agilence  Retail Technology
                         NJ      Edge Therapeutics      Biotechnology
                         NJ                Nistica        Web Hosting

In [86]:
investments.loc[("Closed", 8), ("Name",)]

,Name
State,
CA,CipherMax
CA,Dilithium Networks
CA,Moblyng
CA,SolFocus
CA,Solyndra
FL,Extreme Enterprises
GA,MedShape
NC,Biolex Therapeutics
WA,Cozi Group


In [87]:
# The two lines below are equivalent
investments.xs(key = "NJ", level = 2).head()
investments.xs(key = "NJ", level = "State").head()

Name               Market
Status   Funding Rounds                                      
Acquired 1                         AkaRx        Biotechnology
         1                Aptalis Pharma        Biotechnology
         1                        Cadent             Software
         1               Cancer Genetics  Health And Wellness
         1                     Clacendix           E-Commerce

In [88]:
investments = investments.reset_index()
investments.head()

,Status,Funding Rounds,State,Name,Market
0,Acquired,1,AB,Hallpass Media,Games
1,Acquired,1,AL,EnteGreat,Enterprise Software
2,Acquired,1,AL,Onward Behavioral Health,Biotechnology
3,Acquired,1,AL,Proxsys,Biotechnology
4,Acquired,1,AZ,Envox Group,Public Relations


## 7.8 Summary